airbyte-integrations/connectors/source-zuora/source_zuora/source.py

```diff
@@ -0,0 +1,166 @@
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+#
+
+
+from abc import ABC
+from typing import Any, Dict, Iterable, List, Mapping, MutableMapping, Tuple
+
+from airbyte_cdk import AirbyteLogger
+from airbyte_cdk.models import AirbyteStream
+from airbyte_cdk.sources import AbstractSource
+from airbyte_cdk.sources.streams import Stream
+
+from .zuora_auth import ZuoraAuthenticator
+from .zuora_client import ZoqlExportClient
+from .zuora_errors import ZOQLQueryCannotProcessObject, ZOQLQueryFieldCannotResolve
+
+
+# Main class for Zuora Stream with overriden CDK methods
+class ZuoraStream(Stream, ABC):
+    # Define general primary key
+    primary_key = "id"
+
+    def __init__(self, api: ZoqlExportClient):
+        self.api = api
+
+    def get_cursor_from_schema(self, schema: Dict) -> str:
+        """
+        Get the cursor_field from the stream's schema,
+        If the stream doesn't support 'updateddate', then we use 'createddate'.
+        """
+        return self.alt_cursor_field if self.cursor_field not in schema.keys() else self.cursor_field
+
+    def get_json_schema(self) -> Mapping[str, Any]:
+        """
+        Get the stream's schema from Zuora Object's Data types,
+        """
+        schema = {}
+        schema["properties"] = self.api._zuora_object_to_json_schema(self.name)
+        return schema
+
+    def as_airbyte_stream(self) -> AirbyteStream:
+        """
+        Override this method to set the default_cursor_field to the stream schema.
+        """
+        stream = super().as_airbyte_stream()
+        stream.default_cursor_field = [self.get_cursor_from_schema(stream.json_schema["properties"])]
+        return stream
+
+    def _get_stream_state(self, stream_state: Mapping[str, Any] = None) -> Mapping[str, Any]:
+        """
+        Get the state of the stream for the default cursor_field = updateddate,
+        If the stream doesn't have 'updateddate', then we use 'createddate'.
+        If stream state == None, than we use start_date by default.
+        """
+        return stream_state.get(self.cursor_field, stream_state.get(self.alt_cursor_field)) if stream_state else self.api.start_date
+
+    def read_records(self, stream_state: Mapping[str, Any] = None, **kwargs) -> Iterable[Mapping]:
+        # if stream_state is missing, we will use the start-date from config for a full refresh
+        stream_state = self._get_stream_state(stream_state)
+        try:
+            yield from self.api._get_data_with_date_slice("select", self.name, self.cursor_field, stream_state, self.window_in_days)
+        except ZOQLQueryFieldCannotResolve:
```

Comment:
While we reading the stream, we try to read it using the default cursor_field, which is updateddate.
Since the read method doesn't know which stream has the default cursor_field and which hasn't - we need to switch that dynamically if it's not present in the stream's columns. So, yes, we still have this exception, and we need that to switch the cursor_field correctly.

Reply:
we know what cursor_field base on data from schema. Which is known at the moment when we read_records.

This try is not needed